packages/truffle-reporters/compile/index.js

```diff
@@ -0,0 +1,63 @@
+const colors = require("colors");
+const OS = require("os");
+
+module.exports = {
+  compiledSources(options, sources) {
+    if (!sources) return;
+    const logger = options.logger || console;
+    sources.forEach(source => logger.log("    > " + source));
+  },
+
+  finishJob(options, config) {
+    const { globalConfig } = options;
+    const { compilersInfo } = config;
+    const logger = options.logger || console;
+    if (Object.keys(compilersInfo).length > 0) {
+      logger.log(
+        `${OS.EOL}    > artifacts written to ${
+          globalConfig.contracts_build_directory
+        }`
+      );
+      logger.log(OS.EOL + `Compiled successfully using:` + OS.EOL);
+      for (const name in compilersInfo) {
+        logger.log(`    > ${name}: ${compilersInfo[name].version}`);
+      }
+    } else {
+      logger.log(OS.EOL + `Compilation successful`);
+    }
+    logger.log();
+  },
+
+  initializeListeners(options) {
+    const { emitter } = options;
+    emitter.on("compile:startJob", this.startJob.bind(this, options));
+    emitter.on("compile:finishJob", this.finishJob.bind(this, options));
+    emitter.on("compile:warnings", this.warnings.bind(this, options));
+    emitter.on(
+      "compile:compiledSources",
+      this.compiledSources.bind(this, options)
+    );
+    emitter.on(
+      "compile:nothingToCompile",
+      this.nothingToCompile.bind(this, options)
+    );
```

Comment:
This comment probably belongs someplace more general, but I guess I'll pick on this spot.

This block of code, this whole module, all the similar modules that encapsulate the listener setup and the reporting behavior... having these laid out all in a PR like this is foundational. Thank you for taking the time to write this initial implementation @eggplantzzz; this code and whatever it becomes is going to teach us many lessons.

This feels like our primary resource in designing/validating the entire "subscribe" side of the eventing system. It shows off actual usage when it comes to building a reporter; we can plainly see this new, emerging, external interface to Truffle. If we do our jobs right, there will be thousands or even tens of thousands of files like this one, each of them carving out its own little window into Truffle's operations.

I'll quit with the oratory and summarize: getting this right feels both to be of substantial importance and to pose a significant challenge.

To start with some thoughts about design:

- Encapsulating subscription and handling: yes! I like that these modules define the handling behavior as well as hooking them up as listeners.

- `.on()`s/`.bind()`s for days: intimidating! That's a lot of manual set-up we're requiring, and it's all the same, everywhere!

- These `.on()`s are very far away in the code from the corresponding handler methods! It's hard to look at this collection at a glance and see how things connect. Can we find a way to make an object of event names? Something along the lines of: 
  ```javascript
  {
    "compile:startJob": (data) => { /* ... */ },
    /* ... */
  }
  ```
  Since that could more succinctly represent the important distinctions.

- Magic strings! I strongly urge us to avoid using string literals for these event names. They are risky, difficult to change, and afford no protection against typos.

- Moreover, this whole file relates to `/^compile:/`. Can we make it possible to avoid having to write this everywhere? I was sketching out something like this, but it's difficult to make it not-ugly:
  ```javascript
  module.exports = defineListeners("compile", {
    "startJob": () => {
      this.logger.log("Starting compilation...");
    },

    /* ... */
  });
  ```
  I recognize that this is at odds with the point about magic strings. Was looking for a way to exemplify both in one example, but it's tricky and I've already been at it for an hour.

- Can we avoid passing `options` to every handler? Can we just turn `options` into `this`? (This might add complications the design...)

Anyway, these are just some of my initial thoughts. Hopefully they are useful to the discussion.

Reply:
Another idea (not very considered): what about regex/globbed event listeners?
```javascript
defineListeners("compile", {
  ".*": (data, event) => { console.log("event %s", event) }
});
```

Reply:
Another question: how does initialization work for common things across methods (e.g. `logger`)
```javascript
module.exports = defineListeners({
  namespace: "compile",
  initialize: (config) => {
    this.logger = config.logger || console
  }
}, {
  "startJob": () => {
    this.logger.log("start job")
  },
  /* ... */
});
```